9.2 Домашнее задание Корутины

tunareus opened the pull request:
- Add `item_placeholder.xml` for shimmer layout
- Add shimmer library
- Add `placeholder_background.xml` for item placeholder background
- Update ListFragment
  - Implement shimmer layout
  - Implement load library items and display data
  - Implement error state handling with retry button
  - Implement swipe to delete functionality
- Update DetailFragment
  - Enable save button for creating new item
  - Implement input validation for new item creation
  - Implement save item to viewModel
- Implement `LibraryRepository`
  - Implement getting data from repository
  - Implement adding item to repository
  - Implement deleting item from repository
- Implement `UiState` for view model
- Implement `loadLibraryItems` for view model
- Update `LibraryViewModel`
  - Implement `addNewItem` and `deleteItemAtPosition`
- Add `retry_button_text`, `error_loading_message`, `unknown_error` strings
- Update `MainActivity`
  - Update two pane mode for add item and display detail
  - Fix backstack behaviour for two pane mode
  - Update nav graph for two pane mode

tunareus committed:
feat: Implement data loading with shimmer and error handling

- Add `item_placeholder.xml` for shimmer layout
- Add shimmer library
- Add `placeholder_background.xml` for item placeholder background
- Update ListFragment
  - Implement shimmer layout
  - Implement load library items and display data
  - Implement error state handling with retry button
  - Implement swipe to delete functionality
- Update DetailFragment
  - Enable save button for creating new item
  - Implement input validation for new item creation
  - Implement save item to viewModel
- Implement `LibraryRepository`
  - Implement getting data from repository
  - Implement adding item to repository
  - Implement deleting item from repository
- Implement `UiState` for view model
- Implement `loadLibraryItems` for view model
- Update `LibraryViewModel`
  - Implement `addNewItem` and `deleteItemAtPosition`
- Add `retry_button_text`, `error_loading_message`, `unknown_error` strings
- Update `MainActivity`
  - Update two pane mode for add item and display detail
  - Fix backstack behaviour for two pane mode
  - Update nav graph for two pane mode

tunareus committed:
Refactor Detail and List fragments

- Use `toastMessageFlow` channel instead of `toastMessage` for sending messages.
- Add channel close when ViewModel is destroyed.
- Add error messages for null arguments and validation.
- Add logs.
- Add `isAddingItem` flag in `LibraryViewModel`.
- Add logic to restore state after rotation.
- Add force reset for detail navigation.
- Add navigation from empty detail.
- Prevent redundant start of the shimmer.
- Add two pane mode support for adding items.
- Show detail or add fragment on start.
- Add two-pane mode support.
- Add check to prevent navigation when it is not needed.

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+import java.io.IOException
+
+sealed class UiState {
+    data object Idle : UiState()
```
как будто Idle тут не нужен, начальным состоянием будет Loading

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
-    private val _libraryItems = MutableLiveData<List<LibraryItem>>()
-    val libraryItems: LiveData<List<LibraryItem>> get() = _libraryItems
+    private val repository: LibraryRepository by lazy {
+        Log.d("ViewModel", ">>> Initializing LibraryRepository...")
```
логи хорошо бы поудалять

vladimirlisovskij replied:
и хорошо бы получать репозиторий через конструктор. но думаю про это расскажут в лекции про di

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+    private val _uiState = MutableStateFlow<UiState>(UiState.Idle)
+    val uiState: StateFlow<UiState> = _uiState.asStateFlow()
 
     private val _selectedItem = MutableLiveData<LibraryItem?>()
```
хорошо бы лайв дату заменить на sharedFlow/stateFlow

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+    private val _addItemType = MutableLiveData<String?>(null)
+    val addItemType: LiveData<String?> get() = _addItemType
+
+    private val _toastMessageChannel = Channel<String>(Channel.BUFFERED)
```
а зачем здесь channel? почему не sharedFlow?

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+    private val _toastMessageChannel = Channel<String>(Channel.BUFFERED)
+    val toastMessageFlow = _toastMessageChannel.receiveAsFlow()
+
+    private val _scrollToPosition = MutableSharedFlow<Int>(replay = 0, extraBufferCapacity = 1)
```
а зачем тут буффер?

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+                Log.d("ViewModel", "loadLibraryItems already triggered, skipping call from init.")
+            }
+        } catch (t: Throwable) {
+            Log.e("ViewModel", "Error accessing/initializing repository in init", t)
```
CancellationException не обрабатываем

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+        Log.d("ViewModel", "init called. Initial state is set to Idle.")
+        Log.d("ViewModel", "Accessing repository in init to trigger lazy init...")
+        try {
+            repository
```
лишнее обращение

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+        Log.d("ViewModel", "+++ Entering loadLibraryItems function. forceRetry=$forceRetry, scrollToItemId=$scrollToItemId")
+
+        Log.d("ViewModel", "Current state before launch check: ${_uiState.value::class.java.simpleName}")
+        if (_uiState.value is UiState.Loading && !forceRetry) {
```
думаю тут !forceRetry тут лишний, зачем нам перезапрашивать данные если мы еще не получили ответ на прошлый запрос?

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+
+                    val elapsedTime = System.currentTimeMillis() - coroutineStartTime
+                    Log.d("ViewModel", "Coroutine: Data loaded in $elapsedTime ms.")
+                    if (elapsedTime < 1000) {
+                        val delayNeeded = 1000 - elapsedTime
+                        Log.d("ViewModel", "Coroutine: Delaying for shimmer: $delayNeeded ms")
+                        delay(delayNeeded)
+                    } else {
+                        Log.d("ViewModel", "Coroutine: No shimmer delay needed (elapsed: $elapsedTime ms)")
+                    }
```
а зачем здесь этот блок? проще на уровне репозитория минимальную задержку поставить в секунду

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+        }
+        Log.d("ViewModel", "Preparing to launch new coroutine.")
+
+        try {
```
что мы пытаемся обработать этим блоком try? он выглядит лишним

vladimirlisovskij replied:
да и вложенные try/catch выглядят не очень красиво.  Можешь посмотреть на `runCatching` или на `Result<T>`

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryRepository.kt:
```diff
@@ -0,0 +1,84 @@
+        delay(loadingDelay)
+
+        if (shouldSimulateError()) {
+            throw IOException("Ошибка симуляции: Не удалось загрузить данные из репозитория.")
```
лучше завести свой тип исключения

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+    override fun onCleared() {
+        super.onCleared()
+        Log.d("ViewModel", "onCleared called, cancelling job and closing channel.")
+        loadJob?.cancel()
```
джоба и так завершится вместе со скоупом вьюмодели

vladimirlisovskij commented on app/src/main/res/layout/fragment_list.xml:
```diff
@@ -6,17 +6,87 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:shimmer_auto_start="false">
+
+        <ScrollView
```
хорошо бы показывать шиммеры в ресайклере

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,54 +1,262 @@
+
+
+    private var loadJob: Job? = null
+    private var initialLoadTriggered = false
```
а зачем тут эта переменная? блок init же вызывается только один раз

tunareus committed:
Refactor: Improve data loading and error handling

- Removed redundant logging in ListFragment and DetailFragment.
- Improved error handling in LibraryRepository by introducing a custom `RepositoryLoadException`.
- Refactored `loadLibraryItems` in LibraryViewModel to use `runCatching` for improved error management.
- Replaced `_toastMessageChannel` with `_toastMessage` for simplified toast handling.
- Enhanced navigation logic in MainActivity for smoother transitions between fragments.
- Removed unused LiveData and replaced it with StateFlow and SharedFlow.
- Replaced  `forceRetry` to simple reload `loadLibraryItems`.
- Removed `Idle` state.
- Added a custom exception class `RepositoryLoadException` to manage errors.
- Added time measurement for loading.
- Added a clear loading indicator.
- Added a check for `CancellationException` in order to improve the work of coroutines.

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryRepository.kt:
```diff
@@ -25,16 +27,25 @@ class LibraryRepository {
-        mutex.withLock {
-            internalLibraryItems.toList()
+        if (elapsedTime < 1000) {
+            delay(1000 - elapsedTime)
```
думаю хватило бы и простого delay(1000)

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,262 +1,139 @@
-                        Log.d("ViewModel", "Coroutine: Job cancelled before setting Error state (Throwable).")
+                }
+            }.onFailure { exception ->
+                if (isActive) {
+                    if (exception is CancellationException) {
+                        throw exception
```
тут явно противоречие

vladimirlisovskij replied:
ну и как вариант можно исползовать ensureActive, чтобы убрать лишнее ветвление

vladimirlisovskij replied:
ну и обычно состояние джобы так часто не проверяется тк обычно исключение с отменой прокидывается из вызваных suspend методов

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/LibraryViewModel.kt:
```diff
@@ -1,262 +1,139 @@
-                Log.e("ViewModel", "Error adding item: ${e.message}", e)
-                _toastMessageChannel.send("Ошибка добавления элемента: ${e.message}")
+            }.onSuccess {
+                _toastMessage.tryEmit("Элемент '${item.name}' добавлен.")
```
вот для tryEmit как раз обычно и устанавливают extraBufferCapacity и onBufferOverflow = BufferOverflow.DROP_OLDEST

vladimirlisovskij replied:
а то у меня сейчас тост не показывается

tunareus committed:
Refactor: Improve error handling and loading logic

- Removes unnecessary time measurement in `LibraryRepository`
- Implements `ensureActive` for coroutine cancellation checks
- Cancels previous load jobs in `LibraryViewModel` before starting a new one
- Sets `extraBufferCapacity` and `onBufferOverflow` for `_toastMessage`
